File: ios/ScannerUtils.swift
import Foundation
import MLKitBarcodeScanning

struct Ratio {
    let width: CGFloat
    let height: CGFloat
}

struct Size {
    let width: Int
    let height: Int
}

struct BoundingBox {
    var width: CGFloat
    var height: CGFloat
    var left: CGFloat
    var top: CGFloat
    var right: CGFloat
    var bottom: CGFloat
}

class ScannerUtils {
    static func getOptionsRatio(options: [AnyHashable: Any]?) -> Ratio {
        guard
            let ratio = options?["ratio"] as? [String: Any]
        else {
            return Ratio(width: 1.0, height: 1.0)
        }

        let width = ratio["width"] as? CGFloat ?? 1.0
        let height = ratio["height"] as? CGFloat ?? 1.0
        return Ratio(
            width: min(max(width, 0.0), 1.0),
            height: min(max(height, 0.0), 1.0)
        )
    }

    static func getOptionsBarcodeFormats(options: [AnyHashable: Any]?) -> [Any] {
        if let formats = options?["formats"] as? [Any], !formats.isEmpty {
            return formats
        } else {
            return ["all"]
        }
    }

    static func getSafeBarcodeFormats(formats: [Any]) -> [BarcodeFormat] {
        return formats.compactMap { format -> BarcodeFormat in
            if let format = format as? String {
                switch format {
                    case "code_128": return .code128
                    case "code_39": return .code39
                    case "code_93": return .code93
                    case "codabar": return .codaBar
                    case "ean_13": return .EAN13
                    case "ean_8": return .EAN8
                    case "itf": return .ITF
                    case "upc_e": return .UPCE
                    case "upc_a": return .UPCA
                    case "qr": return .qrCode
                    case "pdf_417": return .PDF417
                    case "aztec": return .aztec
                    case "data_matrix": return .dataMatrix
                    case "all": return .all
                    default: return .all
                }
            } else {
                return .all
            }
        }
    }

    static func getSafeOrientation(orientation: UIImage.Orientation) -> UIImage.Orientation {
        switch orientation {
            case .up: return .up
            case .left: return .right
            case .down: return .down
            case .right: return .left
            default: return .up
        }
    }

    static func getImageSizeWithOrientation(size: Size, orientation: UIImage.Orientation) -> Size {
        switch orientation {
            case .left, .leftMirrored, .right, .rightMirrored:
                return Size(width: size.height, height: size.width)
            default:
                return size
        }
    }

    // This transformation is necessary.
    // Because the default image always comes as landscapeRight.
    // On the Swift side, MLKit does not produce coordinates for a different orientation.
    // This step is done automatically on the Kotlin side.
    static func getFrameRectPortaitCompatible(barcode:Barcode, orientation: UIImage.Orientation, size: Size) -> CGRect {
        let f = barcode.frame

        // By default, the camera captures data in Landscape mode.
        // Here it should always be width > height.
        // imageSize to imageRawSize conversion.
        let imageWidth = CGFloat(max(size.width, size.height))
        let imageHeight = CGFloat(min(size.width, size.height))

        switch orientation {
        case .up:
            return f

        case .left:
            return CGRect(
                x: f.minY,
                y: imageWidth - f.maxX,
                width: f.height,
                height: f.width
            )

        case .right:
            return CGRect(
                x: imageHeight - f.maxY,
                y: f.minX,
                width: f.height,
                height: f.width
            )

        case .down:
            return CGRect(
                x: imageWidth - f.maxX,
                y: imageHeight - f.maxY,
                width: f.width,
                height: f.height
            )

        default:
            return f
        }
    }

    static func getFrameBoxOnRect(rect: CGRect) -> BoundingBox {
        return BoundingBox(
            width: rect.width,
            height: rect.height,
            left: rect.minX,
            top: rect.minY,
            right: rect.maxX,
            bottom: rect.maxY
        )
    }

    static func filterBarcodes(barcodes: [Barcode], size: Size, ratio: Ratio, orientation: UIImage.Orientation) -> [Barcode] {
        let imageWidth = CGFloat(size.width)
        let imageHeight = CGFloat(size.height)

        let scanWidth = ratio.width * imageWidth
        let scanHeight = ratio.height * imageHeight

        let scanLeft = (imageWidth - scanWidth) / 2.0
        let scanTop = (imageHeight - scanHeight) / 2.0
        let scanRight = (imageWidth + scanWidth) / 2.0
        let scanBottom = (imageHeight + scanHeight) / 2.0

        return barcodes.filter { barcode in
            let rect = self.getFrameRectPortaitCompatible(barcode: barcode, orientation: orientation, size: size)
            let box = self.getFrameBoxOnRect(rect: rect)

            return box.left >= scanLeft &&
                   box.top >= scanTop &&
                   box.right <= scanRight &&
                   box.bottom <= scanBottom
        }
    }

    // Normalized for "portrait" mode.
    // If the device is in a different orientation, it should be transformed.
    static func formatBarcode(barcode:Barcode, size: Size, orientation: UIImage.Orientation) -> [String:Any] {
        var map : [String:Any] = [:]
        let rect = self.getFrameRectPortaitCompatible(barcode: barcode, orientation: orientation, size: size)
        let box = self.getFrameBoxOnRect(rect: rect)

        let imageWidth = CGFloat(size.width)
        let imageHeight = CGFloat(size.height)

        let width = box.width
        let height = box.height
        let left = box.left
        let top = box.top
        let right = box.right
        let bottom = box.bottom

        // Raw values
        map["width"] = width
        map["height"] = height
        map["left"] = left
        map["top"] = top
        map["right"] = right
        map["bottom"] = bottom

        // Normalized values
        // Normalizes bounding box to 0–1 range for React Native layout.
        map["leftRatio"] = left / imageWidth
        map["topRatio"] = top / imageHeight
        map["widthRatio"] = width / imageWidth
        map["heightRatio"] = height / imageHeight

        let displayValue = barcode.displayValue
        map["displayValue"] = displayValue
        let rawValue = barcode.rawValue
        map["rawValue"] = rawValue

        let valueType = barcode.valueType
        switch valueType {
            case .wiFi:
                let ssid = barcode.wifi?.ssid
                map["ssid"] = ssid
                let password = barcode.wifi?.password
                map["password"] = password
                let encryptionType = barcode.wifi?.type
                map["encryptionType"] = encryptionType

            case .URL:
                let title = barcode.url!.title
                map["title"] = title
                let url = barcode.url!.url
                map["url"] = url

            default:
                break;
        }

        return map
    }
}

File: android/src/main/java/com/visioncamerabarcodesscanner/ScannerUtils.kt
package com.visioncamerabarcodesscanner

import android.graphics.Rect

import com.facebook.react.bridge.ReadableNativeMap
import com.facebook.react.bridge.WritableNativeMap

import com.google.mlkit.vision.barcode.common.Barcode
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_ALL_FORMATS
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_AZTEC
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODABAR
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODE_128
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODE_39
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_CODE_93
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_DATA_MATRIX
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_EAN_13
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_EAN_8
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_ITF
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_PDF417
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_QR_CODE
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_UPC_A
import com.google.mlkit.vision.barcode.common.Barcode.FORMAT_UPC_E

data class Size(
    val width: Int = 0,
    val height: Int = 0,
)

data class Ratio(
    val width: Float = 1f,
    val height: Float = 1f,
)

data class BoxRatio(
    val leftRatio: Double,
    val topRatio: Double,
    val widthRatio: Double,
    val heightRatio: Double,
)

enum class Orientation (val value: String) {
    PORTRAIT("portrait"),
    LANDSCAPE_RIGHT("landscape-right"),
    PORTRAIT_UPSIDE_DOWN("portrait-upside-down"),
    LANDSCAPE_LEFT("landscape-left");
}

object ScannerUtils {
    fun getOptionsRatio(options: MutableMap<String, Any>?): Ratio {
        val ratio = options?.get("ratio") as? Map<*, *>
        val width = ((ratio?.get("width") as? Number)?.toFloat() ?: 1f).coerceIn(0f, 1f)
        val height = ((ratio?.get("height") as? Number)?.toFloat() ?: 1f).coerceIn(0f, 1f)
        return Ratio(width, height)
    }

    fun getOptionsOrientation(options: MutableMap<String, Any>?): Orientation {
        return when (options?.get("orientation") as? String) {
            "portrait" -> Orientation.PORTRAIT
            "landscape-left" -> Orientation.LANDSCAPE_LEFT
            "portrait-upside-down" -> Orientation.PORTRAIT_UPSIDE_DOWN
            "landscape-right" -> Orientation.LANDSCAPE_RIGHT
            else -> Orientation.PORTRAIT
        }
    }

    fun getOptionsBarcodeFormats(options: MutableMap<String, Any>?): List<*> {
        val formats = options?.get("formats") as? List<*>
        return if (formats.isNullOrEmpty()) {
            listOf("all")
        } else {
            formats
        }
    }

    fun getSafeBarcodeFormats(formats: List<*>): List<Int> {
        return formats.mapNotNull { format ->
            if (format is String) {
                when (format) {
                    "code_128" -> FORMAT_CODE_128
                    "code_39" -> FORMAT_CODE_39
                    "code_93" -> FORMAT_CODE_93
                    "codabar" -> FORMAT_CODABAR
                    "ean_13" -> FORMAT_EAN_13
                    "ean_8" -> FORMAT_EAN_8
                    "itf" -> FORMAT_ITF
                    "upc_e" -> FORMAT_UPC_E
                    "upc_a" -> FORMAT_UPC_A
                    "qr" -> FORMAT_QR_CODE
                    "pdf_417" -> FORMAT_PDF417
                    "aztec" -> FORMAT_AZTEC
                    "data_matrix" -> FORMAT_DATA_MATRIX
                    "all" -> FORMAT_ALL_FORMATS
                    else -> FORMAT_ALL_FORMATS
                }
            } else {
                FORMAT_ALL_FORMATS
            }
        }
    }

    fun getImageSizeByRotation(size: Size, rotationDegrees: Int): Size {
        return when (rotationDegrees) {
            90, 270 -> Size(size.height, size.width)
            else -> size
        }
    }

    private fun getBoxRatioByOrientation(box: Rect, size: Size, orientation: Orientation): BoxRatio {
        val imageWidth = size.width.toDouble()
        val imageHeight = size.height.toDouble()

        // Normalizes bounding box to 0–1 range for React Native layout.
        // Default portrait.
        val leftRatio = box.left / imageWidth
        val topRatio = box.top / imageHeight
        val widthRatio = box.width() / imageWidth
        val heightRatio = box.height() / imageHeight

        return when (orientation) {
            Orientation.PORTRAIT -> BoxRatio(leftRatio, topRatio, widthRatio, heightRatio)
            Orientation.LANDSCAPE_RIGHT -> BoxRatio(
                topRatio,
                1 - leftRatio - widthRatio,
                heightRatio,
                widthRatio
            )
            Orientation.PORTRAIT_UPSIDE_DOWN -> BoxRatio(
                1 - leftRatio - widthRatio,
                1- topRatio - heightRatio,
                widthRatio,
                heightRatio
            )
            Orientation.LANDSCAPE_LEFT -> BoxRatio(
                1 - topRatio - heightRatio,
                leftRatio,
                heightRatio,
                widthRatio
            )
        }
    }

    fun filterBarcodes(barcodes: List<Barcode>, size: Size, ratio: Ratio): List<Barcode> {
        val imageWidth = size.width
        val imageHeight = size.height

        val scanWidth = ratio.width * imageWidth
        val scanHeight = ratio.height * imageHeight

        val scanLeft = ((imageWidth - scanWidth) / 2f).toInt()
        val scanTop = ((imageHeight - scanHeight) / 2f).toInt()
        val scanRight = ((imageWidth + scanWidth) / 2f).toInt()
        val scanBottom = ((imageHeight + scanHeight) / 2f).toInt()

        return barcodes.filter { barcode ->
            val box = barcode.boundingBox
            box != null &&
            box.left >= scanLeft &&
            box.top >= scanTop &&
            box.right <= scanRight &&
            box.bottom <= scanBottom
        }
    }

    fun formatBarcode(barcode: Barcode, size: Size, orientation: Orientation): ReadableNativeMap {
        val map = WritableNativeMap()
        val box = barcode.boundingBox

        if (box != null) {
            val boxRatio = getBoxRatioByOrientation(box, size, orientation)

            // Raw values
            map.putInt("width", box.width())
            map.putInt("height", box.height())
            map.putInt("left", box.left)
            map.putInt("top", box.top)
            map.putInt("right", box.right)
            map.putInt("bottom", box.bottom)

            // Normalized values
            map.putDouble("leftRatio", boxRatio.leftRatio)
            map.putDouble("topRatio", boxRatio.topRatio)
            map.putDouble("widthRatio", boxRatio.widthRatio)
            map.putDouble("heightRatio", boxRatio.heightRatio)
        }

        val rawValue = barcode.rawValue
        map.putString("rawValue", rawValue)
        val displayValue = barcode.displayValue
        map.putString("displayValue", displayValue)
        val valueType = barcode.valueType

        when (valueType) {
            Barcode.TYPE_WIFI -> {
                val ssid = barcode.wifi!!.ssid
                map.putString("ssid", ssid)
                val password = barcode.wifi!!.password
                map.putString("password", password)
                val encryptionType = barcode.wifi!!.encryptionType
                map.putInt("encryptionType", encryptionType)
            }

            Barcode.TYPE_URL -> {
                val title = barcode.url!!.title
                map.putString("title", title)
                val url = barcode.url!!.url
                map.putString("url", url)
            }
        }

        return map
    }
}
